Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -690,184 +687,79 @@ - (void)webView:(WKWebView *)webView runOpenPanelWithParameters:(WKOpenPanelPara
+    case WebExtensionContext::PermissionNotification::GrantedPermissionsWereRemoved:
+        return WKWebExtensionContextGrantedPermissionsWereRemovedNotification;
+    case WebExtensionContext::PermissionNotification::DeniedPermissionsWereRemoved:
+        return WKWebExtensionContextDeniedPermissionsWereRemovedNotification;
+    case WebExtensionContext::PermissionNotification::PermissionMatchPatternsWereGranted:
+        return WKWebExtensionContextPermissionMatchPatternsWereGrantedNotification;
+    case WebExtensionContext::PermissionNotification::PermissionMatchPatternsWereDenied:
+        return WKWebExtensionContextPermissionMatchPatternsWereDeniedNotification;
+    case WebExtensionContext::PermissionNotification::GrantedPermissionMatchPatternsWereRemoved:
+        return WKWebExtensionContextGrantedPermissionMatchPatternsWereRemovedNotification;
+    case WebExtensionContext::PermissionNotification::DeniedPermissionMatchPatternsWereRemoved:
+        return WKWebExtensionContextDeniedPermissionMatchPatternsWereRemovedNotification;
+    case WebExtensionContext::PermissionNotification::Unknown:
+        return nullString();
     }
-
-    if (addedMatchPatterns.isEmpty() && removedMatchPatterns.isEmpty())
-        return;
-
-    removeGrantedPermissionMatchPatterns(addedMatchPatterns, equalityOnly);
-
-    permissionsDidChange(WKWebExtensionContextDeniedPermissionMatchPatternsWereRemovedNotification, removedMatchPatterns);
-    permissionsDidChange(WKWebExtensionContextPermissionMatchPatternsWereDeniedNotification, addedMatchPatterns);
 }
 
-void WebExtensionContext::permissionsDidChange(NSNotificationName notificationName, const PermissionsSet& permissions)
+void WebExtensionContext::permissionsDidChange(WebExtensionContext::PermissionNotification notification, const PermissionsSet& permissions)
 {
     if (permissions.isEmpty())
         return;
 
+    auto notificationName = permissionNotification(notification).createNSString();
+
     if (isLoaded()) {
         RefPtr extensionController = this->extensionController();
         if (!extensionController)
             return;
 
         extensionController->sendToAllProcesses(Messages::WebExtensionContextProxy::UpdateGrantedPermissions(m_grantedPermissions), identifier());
 
-        if (permissions.contains(WKWebExtensionPermissionClipboardWrite)) {
-            bool granted = hasPermission(WKWebExtensionPermissionClipboardWrite);
+        if (permissions.contains(WebExtensionPermission::clipboardWrite())) {
+            bool granted = hasPermission(WebExtensionPermission::clipboardWrite());
 
             enumerateExtensionPages([&](auto& page, bool&) {
                 page.preferences().setJavaScriptCanAccessClipboard(granted);
             });
         }
 
-        if ([notificationName isEqualToString:WKWebExtensionContextPermissionsWereGrantedNotification])
+        if ([notificationName.get() isEqualToString:WKWebExtensionContextPermissionsWereGrantedNotification])
```

Comment:
Comparing to the enum would be better here, I never liked this string comparison!